enhancements/baremetal-config.md

```diff
@@ -0,0 +1,153 @@
+
+### Goals
+
+The goal of this enhancement is to enable the installer to add vital config
+values for a Baremetal IPI deployment to a CR. The MAO will then be able to
+access these values when it kicks off a metal3 cluster.
+
+Update lifecycle for bootimage is an unsolved problem for all platforms. For
+more information, refer to : https://github.com/openshift/os/issues/381.
+
+### Non-Goals
+
+All config values except the RHCOSImageUrl are not expected to change after a
+metal3 cluster is up. Handling changes to this value after install time has
+not been explored and that is a non-goal at this time.
+
+### Proposal
+
+To support configuration of a Baremetal IPI (metal3) deployment, a set of new
+config items are required to be made availabe via a Config Resource. The
+BareMetalPlatformStatus is an ideal fit because it contains configuration
+that is specific to a Baremetal deployment.
+
+This proposal aims to add the following to BareMetalPlatformStatus:
+
+1. ProvisioningInterface : This is the interface name on the underlying
+baremetal server which would be physically connected to the provisioning
+network.
+
+2. ProvisioningIP : This is the IP address used to to bring up a NIC on
+the baremetal server for provisioning. This value should not be in the
+DHCP range and should not be used in the provisioning network for any
+other purpose (should be a free IP address.) It is expected to be provided
+in the format : 10.1.0.3.
+
+3. ProvisioningNetworkCIDR : This is the provisioning network with its
+CIDR. The ProvisioningIP and the ProvisioningDHCPRange are expected to
+be within this network. The value for this config is expected to be
+provided in the format : 10.1.0.0/24.
+
+4. ProvisioningDHCPRange : This is a range of IP addresses from which
+baremetal hosts can be assigned their IP address. This range should be
+provided as two comma seperated IP addresses where the addresses represent
+a contiguous range of IPs that are available to be allocated. It is expected
+to be provided in the format : 10.1.0.10, 10.1.0.100.
+
+5. CacheURL : This is a location where image required to bring up baremetal
+hosts have been previously downloaded for faster image downloads within the
+cluster. This is an optional parameter and is expected to be provided
+in the format : http://111.111.111.1/images
+
+6. RHCOSImageURL : This config is expected to be the location of the
+official RHCOS release image. It is a required parameter and will be used
+to deploy new nodes in the cluster.
+
+### User Stories [optional]
+
+This enhancement will allow users to have their Barametal IPI deployments
+customizable to their specific hardware and network requirements.
+
+This also allows IPI baremetal deployments to take place without manual
+workarounds like creating a ConfigMap for the config (which is the
+current approach we are using to test baremetal deployments.) 
+
+The new config items would be set by the installer and will be used by
```

Comment:
Since the only use-case is _some_ machine-api-operator component, I am inclined towards making these fields available to just that component and not add these to the global configuration.

Reply:
The additions are specific to the baremetal platform in https://github.com/openshift/api/pull/480/ - and this data is required for all deployments when the baremetal platform is selected.

Can you suggest how we'd make this available only to a component?  I'm not entirely clear what you are proposing, and how it differs from using the BareMetalPlatformStatus?

Reply:
> Can you suggest how we'd make this available only to a component?

E.g. [ingress-specific configuration][1] or [the registry-specific config][2] or other [operator.openshift.io][3] stuff.

> ... how it differs from using the BareMetalPlatformStatus?

Mostly in scoping.  It's easier to reason about the impact of config changes if you can scope read access down more tightly.  If the MAO is the only component with creds to read a metal config object, you don't have to worry about other operators or cluster components sneaking in to borrow your settings and give you later trouble when you need to migrate to a different schema.  But if multiple components need access, a central location like Infrastructure is DRYer for cluster admins.

[1]: https://github.com/openshift/api/blob/86def77f6f909e31ffacf73fe63bf5ca830ede83/config/v1/types_ingress.go
[2]: https://github.com/openshift/cluster-image-registry-operator/blob/41484af827099938d629dc6862a9bda4eb1b9a49/pkg/apis/imageregistry/v1/types.go#L97-L170
[3]: https://gcsweb-ci.svc.ci.openshift.org/gcs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-aws-4.3/306/artifacts/e2e-aws/must-gather/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-18b99268c2e0336a6cedb1088b19c5c50601837618578ffb8b2857a36b0618bd/cluster-scoped-resources/operator.openshift.io/